src/query_plan/query_planning_traversal.rs

```diff
@@ -446,7 +470,154 @@ impl<'a> QueryPlanningTraversal<'a> {
+            // Unfortunately, all branches have more than one choices.
+            initial_tree = OpPathTree::new(federated_query_graph.clone(), *root);
+        } else {
+            // Build a tree with the second group's paths.
+            let single_choice_branches: Vec<_> = second_group
+                .iter()
+                .flat_map(|b| &b.0)
+                .flat_map(|cp| cp.flatten())
+                .collect();
+            initial_tree = OpPathTree::from_op_paths(
+                federated_query_graph.clone(),
+                *root,
+                &single_choice_branches,
+            )?;
+            self.updated_dependency_graph(&mut initial_dependency_graph, &initial_tree)?;
+            if first_group.is_empty() {
+                // Well, we have the only possible plan; it's also the best.
+                let cost = self.cost(&mut initial_dependency_graph)?;
+                self.best_plan = BestQueryPlanInfo {
+                    fetch_dependency_graph: initial_dependency_graph,
+                    path_tree: initial_tree.into(),
+                    cost,
+                }
+                .into();
+                return Ok(());
+            }
+        }
+
+        // Build trees from the first group
+        let other_trees: Vec<Vec<Option<Arc<OpPathTree>>>> = first_group
+            .iter()
+            .map(|b| {
+                b.0.iter()
+                    .map(|opt| {
+                        OpPathTree::from_op_paths(
+                            federated_query_graph.clone(),
+                            *root,
+                            &Vec::from_iter(opt.flatten()),
+                        )
+                        .ok()
+                        .map(Arc::new)
+                    })
+                    .collect()
+            })
+            .collect();
+
+        // Implementation of a callback trait to be passed to `generate_all_plans_and_find_best`.
+        struct BestPlanGenerateContext<'a, 'b> {
+            outer_self: &'a mut QueryPlanningTraversal<'b>,
+        }
+
+        impl<'a, 'b> GenerateContext<(FetchDependencyGraph, Arc<OpPathTree>), Arc<OpPathTree>>
+            for BestPlanGenerateContext<'a, 'b>
+        {
+            fn add(
+                &mut self,
+                (plan_graph, plan_tree): &(FetchDependencyGraph, Arc<OpPathTree>),
+                tree: Arc<OpPathTree>,
+            ) -> (FetchDependencyGraph, Arc<OpPathTree>) {
+                let mut updated_graph = plan_graph.clone();
+                let result = self
+                    .outer_self
+                    .updated_dependency_graph(&mut updated_graph, &tree);
+                if result.is_ok() {
+                    let updated_tree = plan_tree.merge(&tree);
+                    (updated_graph, updated_tree)
+                } else {
+                    // Failed to update. Return the original plan.
+                    (updated_graph, plan_tree.clone())
+                }
+            }
```

Comment:
Having read through this impl of the trait's `.add`, I really don't think it needs to be a trait. This add can just live as a function on this struct instead.